cmd/soroban-rpc/internal/db/ledgerentry.go

```diff
@@ -0,0 +1,179 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/jmoiron/sqlx"
+
+	"github.com/stellar/go/xdr"
+)
+
+const (
+	ledgerEntriesTableName = "ledger_entries"
+)
+
+type LedgerEntryReader interface {
+	GetLatestLedgerSequence(ctx context.Context) (uint32, error)
+	NewTx(ctx context.Context) (LedgerEntryReadTx, error)
+}
+
+type LedgerEntryReadTx interface {
+	GetLatestLedgerSequence() (uint32, error)
+	GetLedgerEntry(key xdr.LedgerKey) (bool, xdr.LedgerEntry, error)
+	Done() error
+}
+
+type LedgerEntryWriter interface {
+	UpsertLedgerEntry(key xdr.LedgerKey, entry xdr.LedgerEntry) error
+	DeleteLedgerEntry(key xdr.LedgerKey) error
+}
+
+type ledgerEntryWriter struct {
+	stmtCache *sq.StmtCache
+	buffer    *xdr.EncodingBuffer
+	// nil entries imply deletion
+	keyToEntryBatch map[string]*string
+	maxBatchSize    int
+}
+
+func (l ledgerEntryWriter) UpsertLedgerEntry(key xdr.LedgerKey, entry xdr.LedgerEntry) error {
+	encodedKey, err := encodeLedgerKey(l.buffer, key)
+	if err != nil {
+		return err
+	}
+	// safe since we cast to string right away
+	encodedEntry, err := l.buffer.UnsafeMarshalBinary(&entry)
+	if err != nil {
+		return err
+	}
+	encodedEntryStr := string(encodedEntry)
+	l.keyToEntryBatch[encodedKey] = &encodedEntryStr
+	return l.maybeFlush()
+}
+
+func (l ledgerEntryWriter) DeleteLedgerEntry(key xdr.LedgerKey) error {
+	encodedKey, err := encodeLedgerKey(l.buffer, key)
+	if err != nil {
+		return err
+	}
+	l.keyToEntryBatch[encodedKey] = nil
+	return l.maybeFlush()
+}
+
+func (l ledgerEntryWriter) maybeFlush() error {
+	if len(l.keyToEntryBatch) >= l.maxBatchSize {
+		return l.flush()
+	}
+	return nil
+}
+
+func (l ledgerEntryWriter) flush() error {
+	upsertCount := 0
+	upsertSQL := sq.StatementBuilder.RunWith(l.stmtCache).Replace(ledgerEntriesTableName)
+	var deleteKeys = make([]string, 0, len(l.keyToEntryBatch))
+
+	for key, entry := range l.keyToEntryBatch {
+		if entry != nil {
+			upsertSQL = upsertSQL.Values(key, entry)
+			upsertCount += 1
+		} else {
+			deleteKeys = append(deleteKeys, key)
+		}
+		delete(l.keyToEntryBatch, key)
```

Comment:
hmmm. maybe just clear the map after this for-loop takes place ?

Reply:
honest question: is it cheaper to `delete()` each entry or to garbage collect the full map?

Related: https://stackoverflow.com/questions/13812121/how-to-clear-a-map-in-go

Reply:
I don't know if the map not being cleared out of the GC is a positive or negative here.
Could you add a comment above that line explaining what was the consideration here, and we can revisit that in the future ? ( along with some benchmark that prove that its good/bad )